src/v/bytes/bytes.h

```diff
@@ -16,22 +16,198 @@
+    const_iterator cbegin() const noexcept { return cast(data_.cbegin()); }
+
+    iterator end() noexcept { return cast(data_.end()); }
+    const_iterator end() const noexcept { return cast(data_.end()); }
+    const_iterator cend() const noexcept { return cast(data_.cend()); }
+
+    size_type size() const noexcept { return data_.size(); }
+    bool empty() const noexcept { return data_.empty(); }
+
+    void resize(size_type size) { data_.resize(size); }
+
+    void append(const_pointer s, size_t n) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        data_.append(reinterpret_cast<const char*>(s), n);
+    }
+
+    void append(std::string_view v) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        data_.append(reinterpret_cast<const char*>(v.data()), v.size());
+    }
+
+    bytes& operator+=(const bytes& v) {
+        append(v.data(), v.size());
+        return *this;
+    }
+
+    friend bool operator==(const bytes&, const bytes&) = default;
+
+    friend bool operator<(const bytes& a, const bytes& b) {
+        return a.data_ < b.data_;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const bytes& b);
+
+private:
+    friend class bytes_view;
+
+    static value_type* cast(char* p) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        return reinterpret_cast<value_type*>(p);
+    }
+
+    static const value_type* cast(const char* p) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        return reinterpret_cast<const value_type*>(p);
+    }
+
+    container_type data_;
+};
+
+class bytes_view {
+    using container_type = std::string_view;
+
+public:
+    using value_type = bytes::value_type;
+    using const_pointer = const value_type*;
+
+    bytes_view() = default;
+    bytes_view(const bytes_view&) = default;
+    bytes_view& operator=(const bytes_view&) = default;
+    bytes_view(bytes_view&&) noexcept = default;
+    bytes_view& operator=(bytes_view&&) noexcept = default;
+    ~bytes_view() = default;
+
+    bytes_view(const bytes& bytes)
+      : data_(bytes.data_) {}
+
+    bytes_view(const uint8_t* data, size_t size)
+      : data_(reinterpret_cast<const char*>(data), size) {}
+
+    const_pointer data() const noexcept { return cast(data_.data()); }
+    auto size() const noexcept { return data_.size(); }
+
+    const_pointer begin() const noexcept { return cast(data_.begin()); }
+    const_pointer cbegin() const noexcept { return cast(data_.begin()); }
+
+    const_pointer end() const noexcept { return cast(data_.end()); }
+    const_pointer cend() const noexcept { return cast(data_.end()); }
+
+    friend bool operator==(const bytes_view& a, const bytes_view& b) = default;
+
+    friend bool operator<(const bytes_view& a, const bytes_view& b) {
+        return std::lexicographical_compare(
+          a.begin(), a.end(), b.begin(), b.end());
+    }
```

Comment:
equality is probably ok, but `<` seems sketch since it'd be comparing `char` type and i'm not sure what that means if the expectation is comparison of raw bytes. for example, i'd assume char is signed, so there could be some wrap around. i guess i could just memcmp the raw byte ranges instead.

Reply:
Can we verify with a test?